src/preamble.js

```diff
@@ -1002,12 +987,89 @@ function createWasm(env) {
   };
 #endif
 
+#if WASM && !WASM2JS && MAIN_MODULE
+  // This is similar to what we do in support.js where we resolve
+  // against a local module.
+  var moduleLocal = {};
+  var resolveSymbol = function(sym, type) {
+#if WASM_BACKEND
+    sym = '_' + sym;
+#endif
+    var resolved = Module[sym];
+    if (!resolved)
+      resolved = moduleLocal[sym];
+    
+    // Generate the stub if it doesn't exist
+    if (!resolved) {
+      resolved = moduleLocal[sym] = function() {
+        return Module[sym].apply(null, arguments);
+      };
+    }
+#if ASSERTIONS
+    assert(resolved, 'missing linked ' + type + ' `' + sym + '`. perhaps a side module was not linked in? if this global was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment');
+#endif
+    return resolved;
+  }
+
+  var proxyHandler = {
+    'get': function(obj, prop) {
+      if (prop in obj) {
+        return obj[prop]; // already present
+      }
+
+      // Check if the minification mapping exists first, if not just
+      // use the original name.
+      if (Module["mapping"] && Module["mapping"][prop]) {
+        prop = Module["mapping"][prop];
+      }
+      
+      if (prop.startsWith('g$')) {
+        // a global. the g$ function returns the global address.
+        var name = prop.substr(2); // without g$ prefix
+        return obj[prop] = function() {
+          return resolveSymbol(name, 'global');
+        };
+      }
+
+      if (prop.startsWith('fp$')) {
+        // the fp$ function returns the address (table index) of the function
+        var parts = prop.split('$');
+        assert(parts.length == 3)
+        var name = parts[1];
+        var sig = parts[2];
+        var fp = 0;
+        return obj[prop] = function() {
+          if (!fp) {
+            var f = resolveSymbol(name, 'function');
+            fp = addFunctionWasm(f, sig);
```

Comment:
@kripken , this is where we dynamically generate the stubs for the functions pointed to by the fp$ functions and we subsequently add the function to the Wasm module which results in the OOM crash.

Reply:
But this change should not have increased the number of $fp calls.. the only change should have been that before the addFunctionWasm calls were all static, and they are dynamic.

Perhaps that problem here related to JS scoping rules?  I think perhaps `var fp` is not scoped that way I thought it was.  The idea here was to create new fp for each function, but I think I was maybe wrong about how JS works :(

Reply:
But addFunctionWasm should always be called exactly once for each $fp wasm import.. both before and after this change IIUC